### ISSUE TITLE
How does this compare to `run: corepack enable`?

### ISSUE THREAD

**Issue by mmkal** [pnpm/action-setup#105]
Similar to #30 - I am wondering if there's an advantage to using this vs just running [`corepack enable`](https://nodejs.org/api/corepack.html#enabling-the-feature), which as far as I understand, will instruct node itself to download and install `pnpm`, and will respect the `packageManager` field in package.json.

I'm doing this in my [expect-type](https://github.com/mmkal/expect-type/blob/main/.github/workflows/ci.yml) repo and it seems to be working - is there a downside that you know of?

**Comment by zkochan** [comment]
The advantage is that this installs the "executable" version of pnpm, which is bundled with node.js using `vercel/pkg`. The "executable" version of pnpm is a bit faster and can run on any version of node.js.

**Comment by segevfiner** [comment]
Is this really true? Isn't this a flag https://github.com/pnpm/action-setup/tree/v2#standalone ?

**Comment by paescuj** [comment]
Surely, an advantage over `corepack` is that the action already takes care of store prune, see https://github.com/pnpm/action-setup#use-cache-to-reduce-installation-time.

**Comment by jrr** [comment]
One downside is that corepack is still marked as _experimental_: https://nodejs.org/docs/latest-v20.x/api/corepack.html

**Comment by zkochan** [comment]
@segevfiner yes, it is turned off by default at the moment.

There is nothing wrong with corepack. In our docs we actually recommend corepack for all CIs except Github actions: https://pnpm.io/continuous-integration. But there is nothing wrong with using corepack instead of this action.

**Comment by karlhorky** [comment]
One downside of `pnpm/action-setup` vs `corepack enable` is that it can lead to `ERR_PNPM_BAD_PM_VERSION` failures on `pnpm install` when using [the `packageManager` field in `package.json`](https://nodejs.org/api/packages.html#packagemanager):

1. Configure `packageManager` in `package.json` with an old version eg. `"packageManager": "pnpm@9.0.3"`
2. Use `version: 'latest'` with `pnpm/action-setup`
3. Observe the `ERR_PNPM_BAD_PM_VERSION` error below on `pnpm install` on CI 💥

```
$ pnpm install
 ERR_PNPM_BAD_PM_VERSION  This project is configured to use v9.0.3 of pnpm. Your current pnpm is v9.0.4

If you want to bypass this version check, you can set the "package-manager-strict" configuration to "false" or set the "COREPACK_ENABLE_STRICT" environment variable to "0"
```

## Workaround 1

Switch `pnpm/action-setup` to `corepack enable` to resolve this error:

```diff
-      - uses: pnpm/action-setup@v3
-        with:
-          version: 'latest'
+      - run: corepack enable
```

## Workaround 2

Always keep `packageManager` in `package.json` immediately up to date, eg. using a bot like [Renovate bot](https://docs.renovatebot.com/)

**Comment by segevfiner** [comment]
Doesn't it already have a feature to look at `packageManager` if you don't specify `version`? Check the README, I remember I configured it like this.

**Comment by karlhorky** [comment]
Indeed, I somehow missed this! As per [the `version` docs](https://github.com/pnpm/action-setup#version), to use the `packageManager` version, you can just omit the `version` config:

> `version`
>
> Version of pnpm to install.
>
> Optional when there is a [`packageManager` field in the `package.json`](https://nodejs.org/api/corepack.html).

However, this should probably error in the action when `packageManager` is set and `version` is specified.

I'll create an issue

**Comment by karlhorky** [comment]
Ok I ended up creating a PR for this instead - to throw an error if multiple versions of pnpm are specified:

- https://github.com/pnpm/action-setup/pull/122

**Comment by jasonkuhrt** [comment]
It seems that using core pack does not enable global installs?

```
Run pnpm install -g tsx
 ERR_PNPM_NO_GLOBAL_BIN_DIR  Unable to find the global bin directory

Run "pnpm setup" to create it automatically, or set the global-bin-dir setting, or the PNPM_HOME env variable. The global bin directory should be in the PATH.

```

https://github.com/graffle-js/graffle/actions/runs/11783197937/job/32819852527?pr=1259
